# CHIPsequeando

## What is CHIPsequeando?
CHIPsequeando is an automatic computational workflow specifically designed for the analysis of ChIP-seq data.

## How to install CHIPsequeando?
Download the code from Github to the folder you want. For example: 

```
cd
mkdir ChIPseq
cd ChIPseq
git clone ttps://github.com/AlbertoGonzalezDelgado/CHIPsequeando/ 
```

## Usage
To process ChIP-Seq data run the executable **chip_data_process.sh** with param_input.txt as input
On the folder: param_input you can  find the file where it is neccesary to write this parameters *param_input.txt*:
1. **working_directory:** This parameter specifies the directory where the output folder containing the results of the ChIP-seq data analysis will be generated. For example: *home/biohacker/* 
2. **folder_name:** This parameter specifies the name of the output folder. For example: *work* 
3. **genome:** This parameter specifies the directory that contains the genome file that will be used in the ChIP-seq data analysis. For example: *home/biohacker/experiment/genome/genome.fq.gz* 
4. **annotation:** This parameter specifies the directory that contains the genome annotation that will be used in the ChIP-seq data analysis.For example: *home/biohacker/experiment/annotation/annotation.fq.gz*  
5. **scripts:** This parameter specifies the directory that contains the scripts of CHIPsequeando. For example: *home/ChIPseq/CHIPsequeando* 
6. **number_samples:** This parameter specifies the number of samples 
7. **number_copy:** This parameter specifies the number of copies per each sample.
8. **number_chain_end:** This parameter can take the values **1** when your data is **single-end** or **2** when your data is **paired-end** (two .fastq file per sample). 
9. **chip_x:** This parameter specifies the directory that contains the chips samples. It is neccesary to specified as chips_x as chips samples have to be analysed. For example:
* *Chip_1: home/biohacker/experiment/samples//chip/chip_1.fq.gz* 
* *Chip_2: home/biohacker/experiment/samples//chip/chip_2.fq.gz*
10. **control_x:** his parameter specifies the directory that contains the control or input samples. It is neccesary to specified as controls_x as chips samples have to be analysed. For example:
* *Control_1: home/biohacker/experiment/samples//control/control_1.fq.gz* 
* *Control_1: home/biohacker/experiment/samples//control/control_2.fq.gz*

(Hay que explicar como seria paired end, en el sentido de que seria chip_1 y chip_2 la muestra 1, chip_3 y chip_4 la muestra 2, etc)

Once all parameters have been specified, the next step is to run the scripts. For example: 

```
/home/user/CHIPsequeando/chip_data_process.sh /home/user/CHIPsequeando/param_input/param_file_prr5.txt
```

## Results generated by ChIPsequeando
When ChIPsequeando pipeline is run, the next folders are created in the output directory (specified by the *working_directory* and the *folder_name* parameters): 
1. **annotation:** containing the *annotation.gtf* file.
2. **genome:** containing the *genome.fa* file and the genome index. 
3. **logs:** containing the intermediary and final reports written in *.txt* files. There will be one more files than copies in the experiment. 
4. **results:** containing:
   * *blackboard* file, where each sample processed is noted down.
   * *finishing.txt* file, showing the analysis has been completed.
   * *peaks* files, required by the R script for determining the target genes.
5. **samples:** containing for each copy a *chip* and a *control* folder. Everyone contains the bowtie2 results in *.bam* and *.bam.bai*, the quality analysis results in *.html* and in *.zip* formats, and the sample input in *fq.gz* format
